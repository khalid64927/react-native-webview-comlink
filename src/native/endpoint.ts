import { StringMessageChannel } from 'comlinkjs/messagechanneladapter';
import { WebView, WebViewMessageEvent } from './webview';

export default class WebViewMessageChannel implements StringMessageChannel {
    private webview: WebView;
    private listeners: EventListenerOrEventListenerObject[] = [];

    send(data: string) {
        this.webview.postMessage(data);
    }

    addEventListener(type: string, listener: EventListenerOrEventListenerObject) {
        if (type !== 'message') {
            throw Error(`unsupported event type: ${type}`);
        }

        if (this.listeners.includes(listener)) {
            return;
        }
        this.listeners.push(listener);
    }

    removeEventListener(type: string, listener: EventListenerOrEventListenerObject) {
        if (type !== 'message') {
            throw Error(`unsupported event type: ${type}`);
        }

        const index = this.listeners.indexOf(listener);
        if (index === -1) {
            return;
        }
        this.listeners.splice(index, 1);
    }

    onMessage(event: WebViewMessageEvent) {
        this.listeners.forEach((listener) => {
            if (typeof listener === 'function') {
                listener(event.nativeEvent);
            } else {
                listener.handleEvent(event.nativeEvent);
            }
        });
    }

    setWebview(webview: WebView) {
        this.webview = webview;
    }
}
