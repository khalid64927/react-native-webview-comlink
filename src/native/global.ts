// since react-native global types conflicts with dom types
// we need to exclude dom lib to compile, however, there are some types are
// referenced by comlinkjs, to get it compile, add those types here
interface Window { }

interface EventListenerObject {
    handleEvent(evt: Event): void;
}

interface EventListener {
    (evt: Event): void;
}

declare type EventListenerOrEventListenerObject = EventListener | EventListenerObject;

interface MessageEvent { }

interface MessagePort extends EventTarget {
    onmessage: ((this: MessagePort, ev: MessageEvent) => any) | null;
    onmessageerror: ((this: MessagePort, ev: MessageEvent) => any) | null;
    /**
     * Disconnects the port, so that it is no longer active.
     */
    close(): void;
    /**
     * Posts a message through the channel. Objects listed in transfer are
     * transferred, not just cloned, meaning that they are no longer usable on the sending side.
     * Throws a "DataCloneError" DOMException if
     * transfer contains duplicate objects or port, or if message
     * could not be cloned.
     */
    postMessage(message: any, transfer?: any[]): void;
    /**
     * Begins dispatching messages received on the port.
     */
    start(): void;
    addEventListener(type: string, listener: EventListenerOrEventListenerObject): void;
    removeEventListener(type: string, listener: EventListenerOrEventListenerObject): void;
}

interface MessageChannel {
    readonly port1: MessagePort;
    readonly port2: MessagePort;
}
