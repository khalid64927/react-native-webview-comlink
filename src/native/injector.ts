import { WebView } from 'react-native-webview';
import { script } from './script';

export class Injector {
    private webview: WebView;
    private targetScript: string;
    constructor(name: string, exposedObj: any) {
        // assum exposedObj to be an object with function props
        const exposedTarget = `{${Object.keys(exposedObj)
            .map((key) => `${key}: function(){}`)
            .join(',')}}`;

        this.targetScript = script
            .replace(/\$EXPOSED_NAME/g, `"${name}"`)
            .replace('$EXPOSED_TARGET', exposedTarget);
    }

    inject() {
        if (!this.webview) {
            return;
        }
        this.webview.injectJavaScript(this.targetScript);
    }

    setWebview(webview: WebView) {
        this.webview = webview;
    }
}
