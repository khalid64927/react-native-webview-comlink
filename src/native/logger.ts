export interface Logger {
    (msg: string): void;
}

export function createLogger(debug?: boolean): Logger {
    return (msg) => {
        if (debug) {
            console.log(`[WebViewComlink] ${msg}`);
        }
    }
}
