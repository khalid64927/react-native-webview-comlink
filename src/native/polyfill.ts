// polyfill self for comlinkjs
const globalObj = (function () {
    return this;
})();
if (!globalObj.self) {
    globalObj.self = globalObj;
}
