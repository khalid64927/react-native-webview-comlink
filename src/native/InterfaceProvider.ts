import { MessageHub } from '../common/MessageHub';
import { WebView, WebViewMessageEvent } from 'react-native-webview';
import { Logger } from '../common/logger';

export type isEnabledGetter = () => boolean;

export class InterfaceProvider {
    private webview: WebView;
    private hub: MessageHub;

    constructor(
        name: string,
        rootObj: any,
        private isEnabled: isEnabledGetter,
        private logger: Logger,
    ) {
        this.hub = new MessageHub(name, this);
        Object.keys(rootObj).forEach((key) => {
            const property: unknown = rootObj[key];
            if (typeof property !== 'function') {
                return;
            }
            this.hub.registerReceiver(key, (...args: unknown[]) => {
                return property.apply(rootObj, args);
            });
        });
    }

    postMessage(data: string) {
        this.logger(`sending message to webview: ${data}`);
        if (!this.isEnabled()) {
            this.logger('MessageChannel is disabled, skip sending message');
            return;
        }
        if (!this.webview) {
            this.logger('WebView is missing, skip sending message');
            return;
        }
        (this.webview as any).postMessage(data);
    }

    onMessage(event: WebViewMessageEvent): boolean {
        const { data } = event.nativeEvent;
        if (!this.hub.canHandleMessage(data)) {
            return false;
        }

        this.logger(`received message from webview: ${data}`);
        if (!this.isEnabled()) {
            this.logger('MessageChannel is disabled, drop received message');
            return true;
        }

        this.hub.handleMessage(data);
        return true;
    }

    setWebview(webview: WebView) {
        this.webview = webview;
    }
}
