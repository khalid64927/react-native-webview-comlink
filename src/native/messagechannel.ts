import { StringMessageChannel } from '../common/messagechanneladapter';
import { WebView, WebViewMessageEvent } from 'react-native-webview';
import { Logger } from './logger';

export interface isEnabledGetter {
    (): boolean;
}

export default class WebViewMessageChannel implements StringMessageChannel {
    private webview: WebView;
    private listeners: EventListenerOrEventListenerObject[] = [];

    constructor(private isEnabled: isEnabledGetter, private logger: Logger) {}

    send(data: string) {
        this.logger(`sending message to webview: ${data}`);
        if (!this.isEnabled()) {
            this.logger('MessageChannel is disabled, skip sending message');
            return;
        }
        if (!this.webview) {
            this.logger('WebView is missing, skip sending message');
            return;
        }
        (this.webview as any).postMessage(data);
    }

    addEventListener(type: string, listener: EventListenerOrEventListenerObject) {
        if (type !== 'message') {
            throw Error(`unsupported event type: ${type}`);
        }

        if (this.listeners.includes(listener)) {
            return;
        }
        this.listeners.push(listener);
    }

    removeEventListener(type: string, listener: EventListenerOrEventListenerObject) {
        if (type !== 'message') {
            throw Error(`unsupported event type: ${type}`);
        }

        const index = this.listeners.indexOf(listener);
        if (index === -1) {
            return;
        }
        this.listeners.splice(index, 1);
    }

    onMessage(event: WebViewMessageEvent) {
        this.logger(`received message from webview: ${JSON.stringify(event.nativeEvent.data)}`);
        if (!this.isEnabled()) {
            this.logger('MessageChannel is disabled, drop received message');
            return;
        }

        this.listeners.forEach((listener) => {
            if (typeof listener === 'function') {
                listener(event.nativeEvent);
            } else {
                listener.handleEvent(event.nativeEvent);
            }
        });
    }

    setWebview(webview: WebView) {
        this.webview = webview;
    }
}
