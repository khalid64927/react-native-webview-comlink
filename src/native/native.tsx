import 'message-port-polyfill';
import './polyfill';
import React, { ComponentType, Component, forwardRef } from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { Exposable, expose } from 'comlinkjs';
import { wrap } from '../common/messagechanneladapter';
import {
    WebView,
    WebViewProps,
    WebViewMessageEvent,
    WebViewNavigation,
} from 'react-native-webview';
import { WebViewProgressEvent } from 'react-native-webview/lib/WebViewTypes';
import { WebViewMessageChannel, isEnabledGetter } from './messagechannel';
import { createExposableProxy } from './proxy';
import { Logger, createLogger } from './logger';
import { Injector } from './injector';

type HigherOrderComponentCreator<Props> = (component: ComponentType<Props>) => ComponentType<Props>;

/**
 * withComlinkExpose HOC config
 */
interface Options {
    /**
     * forward ref to the wrapped component, default is `false`
     */
    forwardRef?: boolean;
    /**
     * white list urls where Comlink enabled, default is `null`
     */
    whitelistURLs?: (string | RegExp)[];
    /**
     * for control Comlink enable or disable status, default is `null`
     */
    isEnabled?: isEnabledGetter;
    /**
     * print debug log to console or not, default is `false`
     */
    log?: boolean | Logger;
}

interface RefForwardingProps {
    forwardedRef: any;
}

/**
 * wrap webview component with Comlink support
 * @param rootObj the root object exposed via Comlink
 * @param name the name for the exposed object - window[name] in web side
 * @param config
 */
export function withComlinkExpose<Props extends WebViewProps>(
    rootObj: Exposable,
    name: string,
    options?: Options,
): HigherOrderComponentCreator<Props> {
    const logger: Logger = createLogger(options?.log);
    const whitelistURLs: RegExp[] | null = options?.whitelistURLs?.map((pattern) => {
        if (typeof pattern === 'string') {
            return new RegExp(pattern);
        }
        return pattern;
    });

    return (WrappedComponent: ComponentType<Props>) => {
        class ComponentWithComlinkExpose extends Component<Props & RefForwardingProps> {
            private messageChannel: WebViewMessageChannel;
            private injector: Injector;
            private isCurrentURLInWhitelist: boolean;
            private currentURL: string;

            static displayName: string = `withwithComlinkExpose(${
                WrappedComponent.displayName || WrappedComponent.name
            })`;
            static WrappedComponent: ComponentType<Props> = WrappedComponent;

            constructor(props) {
                super(props);

                this.currentURL = '';
                this.isCurrentURLInWhitelist = true;
                this.onURLUpdated(props.source?.uri);

                // connect and expose the rootObj
                this.messageChannel = new WebViewMessageChannel(this.isEnabled, logger);
                this.injector = new Injector(name, rootObj);
                expose(createExposableProxy(rootObj), wrap(this.messageChannel));
            }

            isEnabled = (): boolean => {
                if (!this.isCurrentURLInWhitelist) {
                    return false;
                }
                if (options?.isEnabled && !options.isEnabled()) {
                    return false;
                }
                return true;
            };

            setWebViewRef = (ref: WebView) => {
                this.messageChannel.setWebview(ref);
                this.injector.setWebview(ref);

                const { forwardedRef } = this.props;
                if (forwardedRef) {
                    if (typeof forwardedRef === 'function') {
                        forwardedRef(ref);
                    } else {
                        forwardedRef.current = ref;
                    }
                }
            };

            onMessage = (event: WebViewMessageEvent) => {
                this.messageChannel.onMessage(event);
                // delegate to event handler
                this.props.onMessage?.(event);
            };

            onLoadStart = (event: Parameters<Props['onLoadStart']>[0]) => {
                const { url } = event.nativeEvent;
                this.onURLUpdated(url);
                if (this.isCurrentURLInWhitelist) {
                    logger(`inject script on page load start`);
                    this.injector.inject();
                }

                // delegate to event handler
                this.props.onLoadStart?.(event);
            };

            onNavigationStateChange = (event: WebViewNavigation) => {
                const { url } = event;
                this.onURLUpdated(url);

                // delegate to event handler
                this.props.onNavigationStateChange?.(event);
            };

            private onURLUpdated(url: string): void {
                if (url === this.currentURL) {
                    return;
                }
                if (!url?.startsWith('http')) {
                    return;
                }

                this.currentURL = url;
                if (whitelistURLs) {
                    // check if the url in whitelist, skip js bridge urls like `react-js-navigation://xxx`
                    this.isCurrentURLInWhitelist = !!whitelistURLs.find((reg) => reg.test(url));
                    logger(`${url} is in whitelist: ${this.isCurrentURLInWhitelist}`);
                }
            }

            render() {
                const { onMessage, onNavigationStateChange, onLoadStart, ...props } = this.props;

                return (
                    <WrappedComponent
                        {...(props as Props)}
                        ref={this.setWebViewRef}
                        onMessage={this.onMessage}
                        onNavigationStateChange={this.onNavigationStateChange}
                        onLoadStart={this.onLoadStart}
                    />
                );
            }
        }

        if (options?.forwardRef) {
            const forwarded = forwardRef((props, ref) => {
                return <ComponentWithComlinkExpose {...(props as Props)} forwardedRef={ref} />;
            }) as any;
            forwarded.displayName = ComponentWithComlinkExpose.displayName;
            forwarded.WrappedComponent = ComponentWithComlinkExpose.WrappedComponent;

            return hoistNonReactStatics(forwarded, WrappedComponent);
        }

        return hoistNonReactStatics(ComponentWithComlinkExpose, WrappedComponent);
    };
}
