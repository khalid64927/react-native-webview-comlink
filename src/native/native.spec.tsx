import React, { Component } from 'react';
import TestRenderer from 'react-test-renderer';

import { withComlinkExpose } from '../native/native';

// TODO: replace with jest.mock
class WebView extends Component {
    public props: any;

    render() {
        return React.createElement('WebView', this.props, this.props.children);
    }

    injectJavaScript() {}
}

it('withComlinkExpose HOC renders correctly', () => {
    const ComlinkExposed = withComlinkExpose({}, 'MyName')(WebView as any);
    const tree = TestRenderer.create(<ComlinkExposed />).toJSON();
    expect(tree).toMatchSnapshot();
});
