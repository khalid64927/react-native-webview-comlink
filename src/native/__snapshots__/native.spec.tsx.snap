// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withComlinkExpose HOC renders correctly 1`] = `
<WebView
  onLoadProgress={[Function]}
  onMessage={[Function]}
  onNavigationStateChange={[Function]}
/>
`;
